File: python/src/table_games/common/cards.py
from enum import Enum
from typing import List
import random

class CSuit(Enum):
    CLUBS = 1
    DIAMONDS = 2
    HEARTS = 3
    SPADES = 4


class CValue(Enum):
    ACE = 1
    TWO = 2
    THREE = 3
    FOUR = 4
    FIVE = 5
    SIX = 6
    SEVEN = 7
    EIGHT = 8
    NINE = 9
    TEN = 10
    JACK = 11
    QUEEN = 12
    KING = 13


class Card:

    def __init__(self, suit: int, value: int) -> None:
        self._suit = suit
        self._value = value

    def __repr__(self) -> str:
        suit_map = {
            CSuit.CLUBS.value:    "♧",
            CSuit.DIAMONDS.value: "♢",
            CSuit.HEARTS.value:   "♡",
            CSuit.SPADES.value:   "♤"
        }

        value_map = {
            CValue.ACE.value:     "A",
            CValue.JACK.value:    "J",
            CValue.QUEEN.value:   "Q",
            CValue.KING.value:    "K"
        }

        suit = suit_map[self._suit]
        value = self._value if self._value not in value_map else value_map[self._value]

        return f"{suit}{value}"


class Deck:

    def __init__(self, cards: List[Card]) -> None:
        self._cards = cards

    def __add__(self, o):
        new_cards = [*self._cards, *o._cards]
        return Deck(new_cards)

    def shuffle(self):
        random.shuffle(self._cards)

    def draw(self):
        return self._cards.pop(0)
    
    def __len__(self):
        return len(self._cards)

    @classmethod
    def Standard(cls):
        cards = []
        
        for suit in CSuit:
            for value in CValue:
                cards.append(Card(suit.value, value.value))

        return Deck(cards)
    

suit_map = {
    'C': CSuit.CLUBS.value,
    'S': CSuit.SPADES.value,
    'H': CSuit.HEARTS.value,
    'D': CSuit.DIAMONDS.value
}
value_map = {
    'A': CValue.ACE.value,
    'K': CValue.KING.value,
    'Q': CValue.QUEEN.value,
    'J': CValue.JACK.value,
    'T': CValue.TEN.value,
    '9': CValue.NINE.value,
    '8': CValue.EIGHT.value,
    '7': CValue.SEVEN.value,
    '6': CValue.SIX.value,
    '5': CValue.FIVE.value,
    '4': CValue.FOUR.value,
    '3': CValue.THREE.value,
    '2': CValue.TWO.value
}

File: rust/table-games/src/common/cards.rs
use std::fmt::{Display, Formatter, Error};
use strum::IntoEnumIterator;
use strum_macros::{EnumIter};
use rand::seq::SliceRandom;
use std::ops::AddAssign;
use std::vec::Vec;

use rand::SeedableRng;
use rand::rngs::StdRng;


#[derive(Debug, EnumIter, Clone, Copy)]
pub enum CardSuit {
    Clubs,
    Diamonds,
    Hearts,
    Spades,
}

impl CardSuit {
    pub fn from_short(short: &str) -> Option<CardSuit> {
        match short {
            "C" => Some(CardSuit::Clubs),
            "D" => Some(CardSuit::Diamonds),
            "H" => Some(CardSuit::Hearts),
            "S" => Some(CardSuit::Spades),
            _ => panic!("Invalid suit: {}", short),
        }
    }
}

#[derive(Debug, EnumIter, Clone, Copy, PartialEq)]
pub enum CardValue {
    Ace = 1,
    Two = 2,
    Three = 3,
    Four = 4,
    Five = 5,
    Six = 6,
    Seven = 7,
    Eight = 8,
    Nine = 9,
    Ten = 10,
    Jack = 11,
    Queen = 12,
    King = 13,
}

impl CardValue {
    pub fn from_short(short: &str) -> Option<CardValue> {
        match short {
            "A" => Some(CardValue::Ace),
            "2" => Some(CardValue::Two),
            "3" => Some(CardValue::Three),
            "4" => Some(CardValue::Four),
            "5" => Some(CardValue::Five),
            "6" => Some(CardValue::Six),
            "7" => Some(CardValue::Seven),
            "8" => Some(CardValue::Eight),
            "9" => Some(CardValue::Nine),
            "T" => Some(CardValue::Ten),
            "J" => Some(CardValue::Jack),
            "Q" => Some(CardValue::Queen),
            "K" => Some(CardValue::King),
            _ => panic!("Invalid value: {}", short),
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub struct Card {
    suit: CardSuit,
    value: CardValue,
}

impl Card {
    pub fn from_short(short: &str) -> Option<Card> {
        let suit = CardSuit::from_short(&short[0..1])?;
        let value = CardValue::from_short(&short[1..2])?;

        Some(Card { suit, value })
    }

    pub fn value(&self) -> CardValue {
        self.value
    }

    pub fn suit(&self) -> CardSuit {
        self.suit
    }
}

impl Display for Card {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
        let suit = match self.suit {
            CardSuit::Clubs => "♣",
            CardSuit::Diamonds => "♦",
            CardSuit::Hearts => "♥",
            CardSuit::Spades => "♠",
        };

        let value = match self.value {
            CardValue::Ace => "A",
            CardValue::Two => "2",
            CardValue::Three => "3",
            CardValue::Four => "4",
            CardValue::Five => "5",
            CardValue::Six => "6",
            CardValue::Seven => "7",
            CardValue::Eight => "8",
            CardValue::Nine => "9",
            CardValue::Ten => "10",
            CardValue::Jack => "J",
            CardValue::Queen => "Q",
            CardValue::King => "K",
        };

        write!(f, "{}{}", value, suit)
    }
}

#[derive(Debug)]
pub struct Deck {
    cards: Vec<Card>,
}

impl Deck {
    pub fn new() -> Deck {
        let mut cards = Vec::new();

        for suit in CardSuit::iter() {
            for value in CardValue::iter() {
                cards.push(Card { suit, value });
            }
        }

        Deck { cards }
    }

    pub fn from_cards(cards: Vec<Card>) -> Deck {
        Deck { cards }
    }

    pub fn shuffle(&mut self) {
        // let mut rng = rand::thread_rng();
        let mut rng = StdRng::from_entropy();
        self.cards.shuffle(&mut rng);
        self.cards.reverse();
        self.cards.shuffle(&mut rng);
    }

    pub fn draw(&mut self) -> Option<Card> {
        self.cards.pop()
    }

    pub fn add(self, rhs: Self) -> Self {
        Deck { cards: [&self.cards[..], &rhs.cards[..]].concat() }
    }

    pub fn len(&self) -> usize {
        self.cards.len()
    }

}

impl AddAssign for Deck {
    fn add_assign(&mut self, rhs: Self) {
        self.cards.extend(rhs.cards);
    }
}

#[test]
fn test_deck() {
    // println!("{}", Deck::new().cards.iter().map(|c| c.to_string()).collect::<Vec<String>>().join(", "));
    assert_eq!(52, Deck::new().len());

    let mut two_decks = Deck::new();
    two_decks += Deck::new();
    assert_eq!(104, two_decks.len());

    two_decks.shuffle();
    assert_eq!(104, two_decks.len());
    
    println!("{}", two_decks.cards.iter().map(|c| c.to_string()).collect::<Vec<String>>().join(", "));
}